src/pip/_internal/req/req_install.py

```diff
@@ -420,12 +420,16 @@ def check_if_exists(self, use_user_site):
         """
         if self.req is None:
             return
+
+        # Canonicalize requirement name to use normalized
+        # names while searching for already installed packages
+        no_marker = Requirement(str(self.req))
+        no_marker.marker = None
+        no_marker.name = canonicalize_name(no_marker.name)
```

Comment:
I wonder if this should use `pkg_resources.safe_name()` instead. Those are functionally equivalent IIUC, but `working_set` uses that, so maybe we should here as well? I don’t know…

Reply:
Hi @uranusjr 

So `pkg_resources.safe_name()` `replaces any runs  of non-alphanumeric/. characters are replaced with a single '-'.` , and `packaging.utils.canonicalize_name` does the same thing as well, but lowercases the string afterwards. 

Also `packaging.utils.canonicalize_name` is directly picked up from [PEP-503](https://www.python.org/dev/peps/pep-0503/#normalized-names) and it even defines the name to be lowercases.

So perhaps `packaging.utils.canonicalize_name` is a better option?

Reply:
We should use whatever `pkg_resources` uses to implement `get_distribution`, so the lookup does not fail when it does not, right? So if it uses `safe_name`, this should as well. (I honestly don’t know, so it would be a good idea to add tests to make sure).

Reply:
Hi @uranusjr 

So I see that https://github.com/pypa/pip/blob/e40d2670facb6cfc815f948c05ca73abb3200f50/src/pip/_vendor/packaging/requirements.py#L98
uses `pyparsing` to parse the string, whi

Apart from this, the only two references I could find for `safe_name` was in [_vendor.Requirements.__init__](https://github.com/pypa/pip/blob/master/src/pip/_vendor/pkg_resources/__init__.py#L3105) and [_internal.req.req_install.InstallRequirement.name](https://github.com/pypa/pip/blob/master/src/pip/_internal/req/req_install.py#L257)

BTW I also see that as per @pradyunsg comment https://github.com/pypa/pip/issues/5021#issuecomment-377768403 , we should be changing the `name` property of `InstallRequirement` to use `canonicalize_name` instead, so perhaps I should modify that property definition instead?

Reply:
There are a lot of moving parts here. `get_distribution()` eventually checks `WorkingSet.by_key`, which is populated with keys from `Distribution.key`, which is in turn `safe_name(...).lower()`. So the technically most fitting way would be to use the same calls here as well: `no_marker.name = safe_name(req.name).lower()`. But as you pointed out, this is functionally the same as `canonicalize_name(req.name)`, so I’d say either would do, but we should add some comments to describe why.

I was referring to the parametrisation when I mentioned the tests.

Reply:
Sure, in that case I will stick to `canonicalize_name` then.

Also I have added a comment stating `Canonicalize requirement name to use normalized names while searching for already installed packages`, should that be enough? Or we need to reword this?

Reply:
I think it’s worth mentioning *why* canonilcaising the name is needed (i.e. `pkg_resources` uses the canonical name to look up packages, but does not process the name passed to `get_distribution()`) for future maintainers stumbling onto this part of the code. It would probably be useful information if we eventually refactor this (#7413).

Reply:
@deveshks Adding *why* we're doing normalization here is a good idea.

Reply:
> @deveshks Adding _why_ we're doing normalization here is a good idea.

Thanks for the reminder. I will do this as well in subsequent commits after using the idea given in https://github.com/pypa/pip/pull/8054/files#r425028035 in my PR.